### ISSUE TITLE
Android Build Failed with Gradle version mismatch issue

### ISSUE THREAD

**Issue by clive107** [flutter/flutter#89368]
## Steps to Reproduce/Issue

* I was able to build the Android app yesterday and a couple of hours ago. Then suddenly I can't build the Android app a few hours ago.
* I did not make any of these changes:
  * Edit `build.gradle`, `app/build.gradle`
  * Add/ remove/ update Flutter packages in `pubspec.yaml`
  * Edit `pubspec.yaml`
  * Upgrade Flutter version
* I am not sure why suddenly unable to build the Android app. I did not update any Flutter packages in the project, it has been working well for months.

## Logs

<details>
<summary>Logs</summary>

```
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Running Gradle task 'assembleDebug'...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugResources'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > Android resource linking failed
     /Users/clive/.gradle/caches/transforms-2/files-2.1/8bd6651f34aabc21707db39f053c2b0c/core-1.7.0-alpha02/res/values/values.xml:105:5-114:25: AAPT: error: resource android:attr/lStar not found.


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 15s
The built failed likely due to AndroidX incompatibilities in a plugin. The tool is about to try using Jetfier to solve the incompatibility.
Building plugin audioplayers...
Running Gradle task 'assembleAarRelease'...
✓ Built build/app/outputs/repo.
Building plugin flutter_secure_storage...
Running Gradle task 'assembleAarRelease'...
✓ Built build/app/outputs/repo.
Building plugin flutter_webview_plugin...
Running Gradle task 'assembleAarRelease'...
✓ Built build/app/outputs/repo.
Building plugin package_info...
Running Gradle task 'assembleAarRelease'...
✓ Built build/app/outputs/repo.
Building plugin path_provider...
Running Gradle task 'assembleAarRelease'...
✓ Built build/app/outputs/repo.
Building plugin qr_code_scanner...
Running Gradle task 'assembleAarRelease'...


FAILURE: Build failed with an exception.

* Where:
Build file '/Users/clive/developer/flutter/.pub-cache/hosted/pub.dartlang.org/qr_code_scanner-0.3.5/android/build.gradle' line: 24

* What went wrong:
A problem occurred evaluating root project 'flutter_qr'.
> Failed to apply plugin [id 'com.android.internal.version-check']
   > Minimum supported Gradle version is 6.5. Current version is 5.6.2. If using the gradle wrapper, try editing the distributionUrl in /Users/clive/developer/flutter/.pub-cache/hosted/pub.dartlang.org/qr_code_scanner-0.3.5/android/gradle/wrapper/gradle-wrapper.properties to gradle-6.5-all.zip

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 654ms

Exception: The plugin qr_code_scanner could not be built due to the issue above.
```

```
[✓] Flutter (Channel stable, 1.22.4, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-MY)
    • Flutter version 1.22.4 at /Users/clive/developer/flutter
    • Framework revision 1aafb3a8b9 (10 months ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/clive/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = /Users/clive/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.8.4

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.59.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.26.0
```

</details>

**Comment by darshankawar** [comment]
@clive107 
The log says 

`Minimum supported Gradle version is 6.5. Current version is 5.6.2. If using the gradle wrapper, try editing the distributionUrl in /Users/clive/developer/flutter/.pub-cache/hosted/pub.dartlang.org/qr_code_scanner-0.3.5/android/gradle/wrapper/gradle-wrapper.properties to gradle-6.5-all.zip`

Try upgrading the gradle version as suggested and see if it works for you.

**Comment by clive107** [comment]
@darshankawar 
* I did not upgrade the `qr_code_scanner` version (been using the same version for months). 
I able to build the Android app for the past few months, and just facing this issue today.
* The reason I don't want to change the `/qr_code_scanner-0.3.5/android/gradle/wrapper/gradle-wrapper.properties` at my local, is because normally I build and release apps using CodeMagic. I need to make it work even build with CodeMagic.

**Comment by darshankawar** [comment]
@clive107 
Since you are on Flutter 1.22 version which is old and you have Android Studio 4.1 which now requires Flutter apps to use newer version of Gradle and hence you may be seeing the issue. 

Also check [this](https://stackoverflow.com/questions/69021225/resource-linking-fails-on-lstar) SO link and see if it works for you.

**Comment by clive107** [comment]
@darshankawar 
I managed to fix the issue with these changes:
1. Update the gradle version in `android/gradle/wrapper/gradle-wrapper.properties`

Before:
```
distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.2-all.zip
```
After:
```
distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-all.zip
```

2. Update the Gradle build tool version in `android/build.gradle`

Before:
```
classpath 'com.android.tools.build:gradle:3.5.0'
```

After:
```
classpath 'com.android.tools.build:gradle:4.1.1'
```

3. Update the compile & target SDK version in `android/app/build.gradle`

Before:
```
compileSdkVersion 29
....
targetSdkVersion 29
```

After:
```
compileSdkVersion 31
....
targetSdkVersion 31
```

---
* After updated the Gradle build tool version, I was able to build APK in debug mode (able to run on Android simulator), but not able to build APK in release mode.
This issue happens on Gradle build tool version 4.x. (Refer to https://github.com/flutter/flutter/issues/58247)
There is one workaround for this issue, which is to add `checkReleaseBuilds false` to `lintOptions` in `android/app/build.gradle`

```
lintOptions {
        ...
        checkReleaseBuilds false
    }
``` 

After making the changes, now I able to build APK in release mode. 

Thank you very much.
